lib/galaxy/datatypes/binary.py

```diff
@@ -1611,6 +1612,37 @@ def display_data(self, trans, dataset, preview=False, filename=None, to_ext=None
         return f"<pre>{repr_}</pre><pre>{rval}</pre>", headers
 
 
+class LudwigModel(Html):
+    """
+    Composite datatype that encloses multiple files for a Ludwig trained model.
+    """
+    composite_type = 'auto_primary_file'
+    file_ext = "ludwig_model"
+
+    def __init__(self, **kwd):
+        super().__init__(**kwd)
+
+        self.add_composite_file('model_hyperparameters.json', description='Model hyperparameters', is_binary=False)
+        self.add_composite_file('model_weights', description='Model weights', is_binary=True)
+        self.add_composite_file('training_set_metadata.json', description='Training set metadata', is_binary=False)
+        self.add_composite_file('training_progress.json', description='Training progress', is_binary=False, optional=True)
+
+    def generate_primary_file(self, dataset=None):
+        rval = ['<html><head><title>Ludwig Model Composite Dataset.</title></head><p/>']
+        rval.append('<div>This model dataset is composed of the following items:<p/><ul>')
+        for composite_name, composite_file in self.get_composite_files(dataset=dataset).items():
+            fn = composite_name
+            opt_text = ''
+            if composite_file.optional:
+                opt_text = ' (optional)'
+            if composite_file.get('description'):
+                rval.append(f"<li><a href=\"{fn}\" type=\"text/plain\">{fn} ({composite_file.get('description')})</a>{opt_text}</li>")
+            else:
+                rval.append(f'<li><a href="{fn}" type="text/plain">{fn}</a>{opt_text}</li>')
```

Comment:
Maybe simplify to something like that? (or something along these lines). That way we have only one parametrized version of the html we are building.

```suggestion
            description = composite_file.get('description')
            link_text = f'{composite_name} ({description})' if description else composite_name
            opt_text = ' (optional)' if composite_file.optional else ''
            rval.append(f'<li><a href="{composite_name}" type="text/plain">{link_text}</a>{opt_text}</li>')
```

Reply:
Super!